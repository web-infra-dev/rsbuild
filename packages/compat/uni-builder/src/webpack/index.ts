import {
  createRsbuild,
  type RsbuildConfig,
  type RsbuildPlugin,
  type RsbuildInstance,
} from '@rsbuild/core';
import type { RsbuildTarget } from '@rsbuild/shared';
import type { UniBuilderWebpackConfig } from '../types';
import type { CreateWebpackBuilderOptions } from '../types';
import { parseCommonConfig } from '../shared/parseCommonConfig';
import { pluginModuleScopes } from './plugins/moduleScopes';
import { pluginStyledComponents } from './plugins/styledComponents';
import { pluginBabel } from './plugins/babel';
import { pluginReact } from './plugins/react';
import { withDefaultConfig } from './defaults';

export async function parseConfig(
  uniBuilderConfig: UniBuilderWebpackConfig,
  cwd: string,
  frameworkConfigPath?: string,
  target?: RsbuildTarget | RsbuildTarget[],
): Promise<{
  rsbuildConfig: RsbuildConfig;
  rsbuildPlugins: RsbuildPlugin[];
}> {
  const { rsbuildConfig, rsbuildPlugins } = await parseCommonConfig<'webpack'>(
    uniBuilderConfig,
    cwd,
    frameworkConfigPath,
    target,
  );

  rsbuildPlugins.push(
    pluginBabel({
      babelLoaderOptions: uniBuilderConfig.tools?.babel,
    }),
  );
  rsbuildPlugins.push(pluginReact());

  if (uniBuilderConfig.tools?.tsLoader) {
    const { pluginTsLoader } = await import('./plugins/tsLoader');
    rsbuildPlugins.push(
      pluginTsLoader(
        uniBuilderConfig.tools.tsLoader,
        uniBuilderConfig.tools.babel,
      ),
    );
  }

  if (uniBuilderConfig.output?.enableAssetManifest) {
    const { pluginManifest } = await import('./plugins/manifest');
    rsbuildPlugins.push(pluginManifest());
  }

  if (uniBuilderConfig.security?.sri) {
    const { pluginSRI } = await import('./plugins/sri');
    rsbuildPlugins.push(pluginSRI(uniBuilderConfig.security?.sri));
  }

  if (uniBuilderConfig.experiments?.lazyCompilation) {
    const { pluginLazyCompilation } = await import('./plugins/lazyCompilation');
    rsbuildPlugins.push(
      pluginLazyCompilation(uniBuilderConfig.experiments?.lazyCompilation),
    );
  }

  rsbuildPlugins.push(
    pluginStyledComponents(uniBuilderConfig.tools?.styledComponents),
  );

  return {
    rsbuildConfig,
    rsbuildPlugins,
  };
}

export async function createWebpackBuilder(
  options: CreateWebpackBuilderOptions,
): Promise<RsbuildInstance> {
  const { rsbuildConfig, rsbuildPlugins } = await parseConfig(
    withDefaultConfig(options.config),
    options.cwd,
    options.frameworkConfigPath,
    options.target,
  );

  const { webpackProvider } = await import('@rsbuild/webpack');
  const rsbuild = await createRsbuild({
    rsbuildConfig,
    provider: webpackProvider,
    target: options.target || 'web',
  });

  rsbuild.addPlugins([
    ...rsbuildPlugins,
    pluginModuleScopes(options.config.source?.moduleScopes),
  ]);

  return rsbuild;
}
