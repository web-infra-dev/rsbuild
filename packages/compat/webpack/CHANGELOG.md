# @rsbuild/webpack

## 0.2.4

### Patch Changes

- @rsbuild/core@0.2.4
- @rsbuild/shared@0.2.4

## 0.2.3

### Patch Changes

- Updated dependencies [44d8087]
  - @rsbuild/core@0.2.3
  - @rsbuild/shared@0.2.3

## 0.2.2

### Patch Changes

- @rsbuild/core@0.2.2
- @rsbuild/shared@0.2.2

## 0.2.1

### Patch Changes

- Updated dependencies [a92997d]
  - @rsbuild/core@0.2.1
  - @rsbuild/shared@0.2.1

## 0.2.0

### Patch Changes

- 42622b45: fix(server-plugin): not relay on modifyRsbuildConfig
- Updated dependencies [c11aa843]
- Updated dependencies [42622b45]
- Updated dependencies [4aa56abd]
- Updated dependencies [04deba66]
  - @rsbuild/core@0.2.0
  - @rsbuild/shared@0.2.0
