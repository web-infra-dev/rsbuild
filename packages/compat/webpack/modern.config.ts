import { configWithMjs } from '../../../scripts/modern.base.config';

export default configWithMjs;
