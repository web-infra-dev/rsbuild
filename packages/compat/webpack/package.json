{
  "name": "@rsbuild/webpack",
  "version": "0.1.6",
  "homepage": "https://rsbuild.dev",
  "repository": {
    "type": "git",
    "url": "https://github.com/web-infra-dev/rsbuild",
    "directory": "packages/compat/webpack"
  },
  "license": "MIT",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    },
    "./plugin-css": {
      "types": "./dist/plugins/css.d.ts",
      "default": "./dist/plugins/css.js"
    }
  },
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "files": [
    "static",
    "compiled",
    "dist"
  ],
  "scripts": {
    "build": "modern build",
    "dev": "modern build --watch"
  },
  "dependencies": {
    "@rsbuild/core": "workspace:*",
    "@rsbuild/shared": "workspace:*",
    "fast-glob": "^3.3.1",
    "globby": "^11.1.0",
    "html-webpack-plugin": "npm:html-rspack-plugin@5.5.7",
    "mini-css-extract-plugin": "2.7.6",
    "postcss": "8.4.31",
    "style-loader": "3.3.3",
    "terser-webpack-plugin": "5.3.9",
    "tsconfig-paths-webpack-plugin": "4.1.0",
    "webpack": "^5.89.0"
  },
  "devDependencies": {
    "@rsbuild/test-helper": "workspace:*",
    "@types/node": "^16",
    "typescript": "^5.3.0"
  },
  "publishConfig": {
    "access": "public",
    "provenance": true,
    "registry": "https://registry.npmjs.org/"
  }
}
