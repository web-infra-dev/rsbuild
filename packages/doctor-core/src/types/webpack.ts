import {
  Compiler,
  Compilation,
  LoaderDefinitionFunction,
  ModuleGraph,
  Dependency,
} from 'webpack';

type GetMapValue<M extends Map<any, any>> = M extends Map<string, infer V>
  ? V
  : never;

export type NormalModuleFactory = ReturnType<
  Compiler['createNormalModuleFactory']
>;
export type SourceMapInput = Parameters<LoaderDefinitionFunction>[1];
export type SourceMap = Exclude<SourceMapInput, string | undefined>;
export type EntryPoint = GetMapValue<Compilation['entrypoints']>;
export type ExportInfo = ReturnType<ModuleGraph['getExportInfo']>;
export type ExportsInfo = ReturnType<ModuleGraph['getExportsInfo']>;

export interface HarmonyImportSpecifierDependency extends Dependency {
  getIds(graph: ModuleGraph): string[];
  name: string;
  userRequest: string;
}
