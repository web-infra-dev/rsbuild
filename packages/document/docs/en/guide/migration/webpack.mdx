# Webpack

This section introduces how to migrate a project using Webpack to Rsbuild.

## Installing Dependencies

First, you need to replace the npm dependencies of Webpack with Rsbuild's dependencies.

import { PackageManagerTabs } from '@theme';

Remove Webpack dependencies:

<PackageManagerTabs command="remove webpack webpack-cli webpack-dev-server" />

Install Rsbuild dependencies:

<PackageManagerTabs command="add @rsbuild/core -D" />

## Updating npm scripts

Next, you need to update the npm scripts in your package.json to use Rsbuild's CLI commands.

```diff title="package.json"
{
  "scripts": {
-   "serve": "webpack serve -c webpack.config.js",
-   "build": "webpack build -c webpack.config.js",
+   "dev": "rsbuild dev",
+   "build": "rsbuild build"
  }
}
```

## Create Configuration File

Create a Rsbuild configuration file `rsbuild.config.ts` in the same directory as package.json, and add the following content:

```ts title="rsbuild.config.ts"
import { defineConfig } from '@rsbuild/core';

export default defineConfig({
  plugins: [],
});
```

## Using Plugins

Rsbuild offers a rich set of plugins that provide out-of-the-box support for common scenarios. You can refer to the [Plugin List](/plugins/list/index) documentation to learn about these plugins.

Taking a React project as an example, let's see how to integrate Rsbuild plugins. First, you can remove some React-related build dependencies that are already built into the Rsbuild React plugin, such as:

- `react-refresh`
- `@babel/preset-react`
- `@pmmmwh/react-refresh-webpack-plugin`

Then see the [React Plugin](/plugins/list/plugin-react) documentation, register it and use it as follows:

```diff
import { defineConfig } from '@rsbuild/core';
+import { pluginReact } from '@rsbuild/plugin-react';

export default {
+  plugins: [pluginReact()],
};
```

## Build Entry

Webpack uses the `entry` field to set the build entry. In Rsbuild, you can use [source.entry](/config/source/entry) to set it.

```ts title="rsbuild.config.ts"
export default {
  source: {
    entry: {
      foo: './src/pages/foo/index.ts',
      bar: './src/pages/bar/index.ts',
    },
  },
};
```

## Configuration Migration

In a Webpack project, there might be some complex `webpack.config.js` configuration files.

After migrating to Rsbuild, most Webpack configurations are built-in and do not require manual configuration, such as output, resolve, module.rules, etc.

For the few Webpack configurations that need to be migrated, you can choose the following options:

- Use the [tools.rspack](/config/tools/rspack) option (Rspack and Webpack configurations are basically equivalent).

```ts title="rsbuild.config.ts"
export default {
  tools: {
    rspack: {
      plugins: [new SomeWebpackPlugin()],
    },
  },
};
```

- Use encapsulated config options in Rsbuild, for example, options for css-loader can be set through [tools.cssLoader](/config/tools/css-loader).

## Validating Results

After completing the above steps, you have completed the basic migration from Webpack to Rsbuild. You can now run the `npm run serve` command to try starting the dev server.

If you encounter any issues during the build process, please debug according to the error log, or check the Webpack configuration to see if there are any necessary configurations that have not been migrated to Rsbuild.

## Cleaning Up Dependencies

Since Rsbuild has built-in some common loaders and plugins, you can remove the following dependencies, which will significantly improve the dependency installation speed of the project:

- css-loader
- less-loader
- sass-loader
- babel-loader
- style-loader
- postcss-loader
- html-webpack-plugin
- mini-css-extract-plugin
- autoprefixer
- @babel/core
- @babel/preset-env
- @babel/preset-typescript
- @babel/runtime
- ...

:::tip
The above only lists some of the common dependencies that can be removed. In actual Webpack projects, there may be many other dependencies, please handle them as appropriate.
:::

## Contents Supplement

The current document only covers part of the migration process. If you find suitable content to add, feel free to contribute to the documentation via pull request 🤝.

> The documentation for rsbuild can be found in the [rsbuild/packages/document](https://github.com/web-infra-dev/rsbuild/tree/main/packages/document) directory.
