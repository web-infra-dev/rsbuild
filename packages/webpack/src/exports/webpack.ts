import webpack from 'webpack';

export default webpack;
