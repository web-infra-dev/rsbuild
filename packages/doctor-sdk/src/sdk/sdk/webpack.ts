import { DevToolError } from '@rsbuild/doctor-utils/error';
import { Common, Manifest, SDK } from '@rsbuild/doctor-types';
import { File } from '@rsbuild/doctor-utils/build';
import { RawSourceMap, SourceMapConsumer } from 'source-map';
import { ModuleGraph, ChunkGraph, PackageGraph } from '../../graph';
import { debug } from '@rsbuild/doctor-utils/logger';
import { DoctorServer } from '../server';
import { DoctorFakeServer } from '../server/fakeServer';
import { DoctorWebpackSDKOptions } from './types';
import { SDKCore } from './core';

export class DoctorWebpackSDK<
    T extends DoctorWebpackSDKOptions = DoctorWebpackSDKOptions,
  >
  extends SDKCore<T>
  implements SDK.DoctorBuilderSDKInstance
{
  public server: DoctorServer;

  public extraConfig: SDK.SDKOptionsType | undefined;

  private type: SDK.ToDataType;

  private _summary: SDK.SummaryData = { costs: [] };

  private _configs: SDK.ConfigData = [];

  private _errors: DevToolError[] = [];

  private _loader: SDK.LoaderData = [];

  private _resolver: SDK.ResolverData = [];

  private _plugin: SDK.PluginData = {};

  private _moduleGraph = new ModuleGraph();

  private _chunkGraph = new ChunkGraph();

  private _rawSourceMapCache = new Map<string, RawSourceMap>();

  private _sourceMap = new Map<string, SourceMapConsumer>();

  private _packageGraph!: SDK.PackageGraphInstance;

  private _tileReportHtml: string | undefined;

  constructor(options: T) {
    super(options);
    this.server = options.noServer
      ? new DoctorFakeServer(this, undefined)
      : new DoctorServer(this, options.port);
    this.type = options.type || SDK.ToDataType.Normal;
    this.extraConfig = options.config;
  }

  async bootstrap() {
    debug(() => `${Date.now()}`, '[DoctorWebpackSDK][bootstrap start]');
    this.server && (await this.server.bootstrap());
    await super.bootstrap();
    debug(
      () => `${Date.now()} ${this.server.origin}`,
      '[DoctorWebpackSDK][bootstrap end]',
    );
  }

  async dispose() {
    debug(() => `${Date.now()}`, '[DoctorWebpackSDK][dispose start]');
    this.server && (await this.server.dispose());
    await super.dispose();
    debug(() => `${Date.now()}`, '[DoctorWebpackSDK][dispose end]');
  }

  async applyErrorFix(id: number) {
    const { _errors: errors } = this;
    const error = errors.find((err) => err.id === id);

    if (!error || !error.path || !error.fixData || error.fixData.isFixed) {
      return;
    }

    const { path: filePath, fixData } = error;
    const sameFileErrors = errors.filter(
      (item) => item.path === filePath && item !== error,
    );
    let content = (await File.fse.readFile(filePath, 'utf-8')).toString();

    // Application of current modification.
    const startTxt = content.substring(0, fixData.start);
    const endTxt = content.substring(fixData.end, content.length);
    const offset =
      (fixData.newText ?? '').length - (fixData.end - fixData.start);

    // Modified text.
    content = startTxt + fixData.newText + endTxt;

    // The remaining modifications of the same document need to be recalculated.
    for (const other of sameFileErrors) {
      const { fixData: otherFixData } = other;

      if (!otherFixData) {
        continue;
      }

      // After the modified text, the offset needs to be corrected.
      if (otherFixData.start >= fixData.end) {
        otherFixData.start += offset;
        otherFixData.end += offset;
      }
    }

    // Modify the write to the hard disk.
    await File.fse.writeFile(filePath, content);
  }

  clear() {
    this._errors = [];
    this._loader = [];
    this._resolver = [];
    this._plugin = {};
    this._moduleGraph = new ModuleGraph();
    this._chunkGraph = new ChunkGraph();
  }

  clearSourceMapCache(): void {
    this._rawSourceMapCache = new Map();
    this._sourceMap = new Map();
  }

  async getSourceMap(file: string): Promise<SourceMapConsumer | undefined> {
    const { _sourceMap: sourceMap, _rawSourceMapCache: rawMap } = this;

    if (sourceMap.has(file)) {
      return sourceMap.get(file)!;
    }

    const rawData = rawMap.get(file);

    // There is no data, or illegal data.
    if (
      !rawData ||
      rawData.version < 0 ||
      !rawData.sourcesContent?.[0] ||
      !rawData.mappings
    ) {
      return Promise.resolve(undefined);
    }

    try {
      const result = await new SourceMapConsumer(rawData);
      sourceMap.set(file, result);
      return result;
    } catch (e) {
      // TODO: Specific errors need to be checked.
      return Promise.resolve(undefined);
    }
  }

  reportSourceMap(data: RawSourceMap): void {
    this._rawSourceMapCache.set(data.file, data);
  }

  reportConfiguration(config: SDK.ConfigData[0]) {
    this._configs.push(config);
    this.onDataReport();
  }

  reportError(errors: Error[]) {
    errors.forEach((item) => {
      this._errors.push(
        DevToolError.from(item, {
          code: this.name,
        }),
      );
    });
    this.onDataReport();
  }

  reportLoader(data: SDK.LoaderData) {
    data.forEach((item) => {
      // find by resource.path
      let match = this._loader.find(
        (e) => e.resource.path === item.resource.path,
      );
      if (match) {
        match.loaders.push(...item.loaders);
      } else {
        match = item;
        this._loader.push(item);
      }

      match.loaders.sort((a, b) => {
        // default sort by call timestamp
        if (a.startAt !== b.startAt) {
          return a.startAt - b.startAt;
        }

        if (a.isPitch) {
          if (b.isPitch) {
            return a.loaderIndex - b.loaderIndex;
          }
          // [a, b]
          return -1;
        }

        if (b.isPitch) {
          // [b, a]
          return 1;
        }

        return b.loaderIndex - a.loaderIndex;
      });
    });
    this.onDataReport();
  }

  reportResolver(data: SDK.ResolverData): void {
    data.forEach((item) => this._resolver.push(item));
    this.onDataReport();
  }

  reportPlugin(data: SDK.PluginData): void {
    Object.keys(data).forEach((hook) => {
      if (!this._plugin[hook]) {
        this._plugin[hook] = data[hook];
      } else {
        data[hook].forEach((item) => {
          this._plugin[hook].push(item);
        });
      }
    });
    this.onDataReport();
  }

  reportModuleGraph(data: ModuleGraph): void {
    debug(() => `data size: ${data.size()}`, '[SDK.reportModuleGraph][start]');
    this._moduleGraph.fromInstance(data);
    this.createPackageGraph();
    this.onDataReport();
    debug(
      () => `sdk._moduleGraph size: ${this._moduleGraph.size()}`,
      '[SDK reportModuleGraph][end]',
    );
  }

  reportPackageGraph(data: PackageGraph): void {
    debug(() => '[SDK.reportPackageGraph][start]');
    if (!this._packageGraph) {
      this._packageGraph = data;
    }
    this.onDataReport();
    debug(
      () => `sdk._moduleGraph size: ${this._moduleGraph.size()}`,
      '[SDK reportPackageGraph][end]',
    );
  }

  reportChunkGraph(data: ChunkGraph): void {
    this._chunkGraph.addAsset(...data.getAssets());
    this._chunkGraph.addChunk(...data.getChunks());
    this._chunkGraph.addEntryPoint(...data.getEntryPoints());
    this.onDataReport();
  }

  reportSummaryData(part: Partial<SDK.SummaryData>): void {
    const keys = ['costs'] as Array<keyof SDK.SummaryData>;

    for (const key of keys) {
      const v = part[key];
      if (!v) continue;
      if (typeof v === 'object') {
        if (Array.isArray(v)) {
          (this._summary[key] as unknown[]) = [
            ...((this._summary[key] as Array<any>) || []),
            ...v,
          ];
        } else {
          (this._summary[key] as Common.PlainObject) = {
            ...((this._summary[key] as Common.PlainObject) || {}),
            ...(v as Common.PlainObject),
          };
        }
      } else {
        (this._summary[key] as unknown) = v;
      }
    }

    this.onDataReport();
  }

  reportTileHtml(tileReportHtml: string): void {
    this._tileReportHtml = tileReportHtml;
  }

  createPackageGraph() {
    debug(
      () => `sdk._moduleGraph size: ${this._moduleGraph.size()}`,
      '[SDK.createPackageGraph][start]',
    );
    if (!this._packageGraph) {
      const pkgGraph = PackageGraph.fromModuleGraph(
        this._moduleGraph,
        this.root,
        (path: string) => {
          try {
            const exists = File.fse.existsSync(path);
            if (exists) {
              debug(
                () =>
                  `sdk.PackageGraph package.json exists: ${exists}, path: ${path}`,
                '[SDK.createPackageGraph][load]',
              );
              return File.fse.readJSONSync(path);
            }
          } catch (error) {
            const { message, stack } = error as Error;
            debug(
              () =>
                `sdk.createPackageGraph error, path: ${path}, error message: ${
                  stack || message
                }`,
              '[SDK.createPackageGraph][error]',
            );
          }
        },
      );
      this._packageGraph = pkgGraph;
      debug(
        () =>
          `sdk._packageGraph packages: ${
            this._packageGraph.getPackages().length
          }`,
        '[SDK.createPackageGraph][end]',
      );
    }
  }

  public writeStore(options?: SDK.WriteStoreOptionsType) {
    debug(() => `sdk.writeStore has run.`, '[SDK.writeStore][end]');
    return this.saveManifest(this.getStoreData(), options || {});
  }

  public getStoreData(): SDK.BuilderStoreData {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const ctx = this;
    return {
      get hash() {
        return ctx.hash;
      },
      get root() {
        return ctx.root;
      },
      get envinfo() {
        return ctx._envinfo;
      },
      get pid() {
        return ctx.pid;
      },
      get errors() {
        return ctx._errors.map((err) => err.toData());
      },
      get configs() {
        return ctx._configs.slice();
      },
      get summary() {
        return { ...ctx._summary };
      },
      get resolver() {
        return ctx._resolver.slice();
      },
      get loader() {
        return ctx._loader.slice();
      },
      get moduleGraph() {
        return ctx._moduleGraph.toData({
          contextPath: ctx._configs?.[0]?.config?.context || '',
        });
      },
      get chunkGraph() {
        return ctx._chunkGraph.toData(ctx.type);
      },
      get moduleCodeMap() {
        return ctx._moduleGraph.toCodeData(ctx.type);
      },
      get plugin() {
        return { ...ctx._plugin };
      },
      get packageGraph() {
        return ctx._packageGraph
          ? ctx._packageGraph.toData()
          : {
              packages: [],
              dependencies: [],
            };
      },
      get otherReports() {
        return { tileReportHtml: ctx._tileReportHtml || '' };
      },
    };
  }

  public getManifestData(): Manifest.DoctorManifestWithShardingFiles {
    const dataValue = this.getStoreData();
    const data: Manifest.DoctorManifestWithShardingFiles = {
      client: {
        enableRoutes: this.getClientRoutes(),
      },
      data: Object.keys(dataValue).reduce((t, e) => {
        const _e = e as keyof SDK.BuilderStoreData;
        if (dataValue[_e] && typeof dataValue[_e] === 'object') {
          t[e] = [
            `${this.server.origin}${SDK.ServerAPI.API.LoadDataByKey}/${e}`,
          ];
        } else {
          t[e] = dataValue[_e];
        }

        return t;
      }, {} as Common.PlainObject) as unknown as Manifest.DoctorManifestWithShardingFiles['data'],
      __LOCAL__SERVER__: true,
      __SOCKET__URL__: this.server.socketUrl,
    };

    return data;
  }

  public getRuleContext(
    _options: SDK.RuntimeContextOptions,
  ): SDK.RuntimeContext {
    this.createPackageGraph();

    return {
      root: this.root,
      errors: this._errors.slice(),
      configs: this._configs.slice(),
      moduleGraph: this._moduleGraph,
      chunkGraph: this._chunkGraph,
      packageGraph: this._packageGraph,
      loader: this._loader.slice(),
      otherReports: { tileReportHtml: this._tileReportHtml || '' },
    };
  }

  public onDataReport(): void | Promise<void> {
    this.server.broadcast();
  }
}
