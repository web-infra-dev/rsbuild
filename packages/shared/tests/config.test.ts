import webpack from 'webpack';
import {
  getExtensions,
  stringifyConfig,
  pickRsbuildConfig,
} from '../src/config';

it('should get default extensions correctly', async () => {
  expect(getExtensions()).toEqual(['.js', '.jsx', '.mjs', '.json']);
});

it('should get extensions with ts', async () => {
  expect(
    getExtensions({
      isTsProject: true,
    }),
  ).toEqual(['.ts', '.tsx', '.js', '.jsx', '.mjs', '.json']);
});

it('should get extensions with prefix', async () => {
  expect(
    getExtensions({
      resolveExtensionPrefix: '.web',
    }),
  ).toEqual([
    '.web.js',
    '.js',
    '.web.jsx',
    '.jsx',
    '.web.mjs',
    '.mjs',
    '.web.json',
    '.json',
  ]);
});

it('should get extensions with prefix by target', async () => {
  expect(
    getExtensions({
      target: 'node',
      resolveExtensionPrefix: {
        web: '.web',
        node: '.node',
      },
    }),
  ).toEqual([
    '.node.js',
    '.js',
    '.node.jsx',
    '.jsx',
    '.node.mjs',
    '.mjs',
    '.node.json',
    '.json',
  ]);
});

describe('stringifyConfig', () => {
  it('should stringify webpack config correctly', async () => {
    const { DefinePlugin } = webpack;
    const webpackConfig = {
      mode: 'development',
      plugins: [new DefinePlugin({ foo: 'bar' })],
    };

    expect(await stringifyConfig(webpackConfig)).toMatchSnapshot();
  });

  it('should stringify webpack config with verbose option correctly', async () => {
    const { DefinePlugin } = webpack;
    const webpackConfig = {
      mode: 'development',
      plugins: [
        new DefinePlugin({
          foo: 'bar',
          baz() {
            const a = 1;
            return a;
          },
        }),
      ],
    };

    expect(await stringifyConfig(webpackConfig, true)).toMatchSnapshot();
  });

  it('should stringify Rsbuild config correctly', async () => {
    const rsbuildConfig = {
      tools: {
        webpackChain(chain: any) {
          chain.devtool('eval');
        },
      },
    };

    expect(await stringifyConfig(rsbuildConfig)).toMatchSnapshot();
  });
});

describe('pickRsbuildConfig', () => {
  it('should pick correct keys from Rsbuild config', () => {
    const rsbuildConfig = {
      dev: {},
      html: {},
      tools: {},
      source: {},
      output: {},
      security: {},
      performance: {},
      extraKey: 'extraValue',
    };

    const result = pickRsbuildConfig(rsbuildConfig);

    expect(result).toEqual({
      dev: {},
      html: {},
      tools: {},
      source: {},
      output: {},
      security: {},
      performance: {},
    });
  });

  it('should return empty object when Rsbuild config is empty', () => {
    const rsbuildConfig = {};
    const result = pickRsbuildConfig(rsbuildConfig);
    expect(result).toEqual({});
  });
});
