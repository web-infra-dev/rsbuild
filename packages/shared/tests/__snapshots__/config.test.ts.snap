// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`stringifyConfig > should stringify Rsbuild config correctly 1`] = `
"{
  tools: {
    webpackChain(chain) {
      chain.devtool("eval");
    }
  }
}"
`;

exports[`stringifyConfig > should stringify webpack config correctly 1`] = `
"{
  mode: 'development',
  plugins: [
    {
      definitions: {
        foo: 'bar'
      }
    }
  ]
}"
`;

exports[`stringifyConfig > should stringify webpack config with verbose option correctly 1`] = `
"{
  mode: 'development',
  plugins: [
    {
      definitions: {
        foo: 'bar',
        baz() {
          const a = 1;
          return a;
        }
      }
    }
  ]
}"
`;
